en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
-# 创建新的Op
+# Build New Operator
 
-## 背景
-本文档服务于那些需要自定义实现C++ Op的用户，通常来说，这来源于如下需求
+## Background 
 
-- 自定义Op难以通过OneFlow已有的Op在Python前端组合搭配生成。
-- 自定义Op可以通过OneFlow已有Op在Python前端组合搭配生成，但却满足不了性能需求。
-- 用户想要手动对kernel进行融合fuse，以满足某些特定的需求。
+### What is a Custom Op
 
-除上述需求之外的需求，我们推荐您在Python前端使用现有Op组合得到想要的Op。
+OneFlow abstracts all kinds of data processing into  op (operator). Op is acts on the input tensor and writes the result of the operation to the output tensor. OneFlow provides comprehensive ops and can be found in the [ops directory](https://github.com/Oneflow-Inc/oneflow/tree/master/oneflow/python/ops).
 
+When OneFlow's existing Python operators are not sufficient to build a neural network or when Python operators do not meet performance requirements. You can use C++ to develop custom op in OneFlow.
 
+OneFlow provides a system which you can create custom op and register it in OneFlow then use custom op in Python.
 
-对于的确需要手动实现C++ Op的用户，你需要如下步骤来让自定义Op正常工作：
+The following diagram demonstrates the registration system for a custom op in OneFlow.
 
-1. 注册Op的定义： Op的定义独立于Op的实现（即Kernel），用来描述Op的功能性。通常包括Op的名称，Op的输入和输出，Op的配置属性和一些必要的用于推导Tensor的shape和data type的函数。
-2. 用C++实现Op对应的Kernel： Kernel用来描述Op的详细计算过程。对于一个Op来说，可能会对应多个Kernel。
-3. 书写对应的Python前端：因为OneFlow的网络构建是基于Python去书写的，所以我们需要在Python前端去书写少量代码来封装前两步所写的C++的代码。
-4. （可选）注册Op对应的后向构图函数：如果训练网络中需要Op的反向，那么我们还需要去写一个函数来告诉OneFlow如何在去构建该Op对应的后向计算过程。
-5. 测试Op：上述步骤全部执行完毕后，我们还需要对Op进行测试以保证Op的正确性，具体步骤见后续。
+![OneFlow UserOp Existing System](imgs/oneflow_system_userop.png)
 
+In the OneFlow framework, there are three types of registries associated with custom op.
 
-## 定义新的Op
+* `OpGradRegistry`：Manage gradient registration for automatic gradient calculation in converse digraph.
 
-我们以`Relu`为例，来走一遍Op注册的流程，构建一个cpp文件`relu.cpp`用于注册Op。
+* `OpRegistry`：Managing op registrations for generating forward digraph and building`Task Graph`.
 
-``` cpp
+* `OpKernelRegistry`：Manage kernel registrations for executing user-written kernel logic when running.
```

Comment:
Manage kernel registrations for executing user-written kernel logic when running.
->
Manage kernel registrations for peform user logic at runtime.